cip-multiout-reverse.md

```diff
@@ -0,0 +1,138 @@
+
+For example, if the service process a Multi-Out with transaction ID `TransactionA` and accounts `AccountA`, `AccountB` and `AccountC` are paid an amount, and then processes a Multi-Out Same with transaction ID `TransactionB` where accounts `AccountA` and `AccountD` are paid, it would add the rows:
+
+| recipient_id | transaction_id |
+|-|-|
+| AccountA | TransactionA |
+| AccountB | TransactionA |
+| AccountC | TransactionA |
+| AccountA | TransactionB |
+| AccountD | TransactionB |
+
+2. When a user requests their transaction list, search the table for all rows where `recipient` is equal to the user's account ID, which would in this example for `AccountA` return:
+
+| recipient_id | transaction_id |
+|-|-|
+| AccountA | TransactionA |
+| AccountA | TransactionB |
+
+3. Lookup the transactions found and add them to the list of regular transactions obtained from the transactions table in the database.
+
+4. When displaying the list of transactions, display the incoming amount as the amount that `AccountA` was paid in the Multi-Out (Same) transaction. There will only be one entry in the multi-out recipients as accounts cannot be sent to twice in the same Multi-Out (Same) transaction.
+
+### Impact on database storage
+
+Each row in the new table consumes 16 bytes, and there is one per recipient per transaction.
+
+Therefore the maximum amount of additional space consumed per Multi-Out transaction is `16*64 = 1024` bytes, which is `1024/1201 = 85%` more than was previously consumed per Multi-Out transaction.
+
+Additionally, the maximum amount of additional space consumed per Multi-Out Same transaction is `16*128 = 2048` bytes, which is `2048/1201 = 171%` more than was previously consumed per Multi-Out transaction.
+
+Additionally, the new space consumption decrease (compared to not having Multi-Out transactions) is now `(11264-(1201+1024))/11264 = 80.2%` (down from `89.3%` without the new table).
+
+Additionally, the new space consumption decrease (compared to not having Multi-Out Same transactions) is now `(22528-(1201+2048))/22528 = 85.6%` (down from `94.7%` without the new table).
+
+Therefore, I am satisfied that objective #4 is fulfilled, as we are taking a relatively minor hit in storage consumption for the added benefit of having multi-out reverse lookup (especially considering that Multi-Outs are used considerably more than Multi-Out Sames as the network pools are constantly using them, and they are affected less by this change)
+
+## Compatibility
+
+This will affect users of the `getAccountTransactions` and `getAccountTransactionIDs` API calls, so we should add a boolean flag to the request parameters to "Include Incoming Multi-Outs" which will default to false, as otherwise clients using the API that have not updated will be confused by the fact that the account is not the recipient of the transaction and by the fact that the transaction amount is not the amount sent to the account.
```

Comment:
This is awesome, will allow me to remove some hax from BurstAlerts 👍

Reply:
liked that, too